File: Problem_3 Spiral Matrix.py
# // Time Complexity : O(mn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or len(matrix) == 0:
            return []
        top = 0
        bottom = len(matrix) - 1
        right = len(matrix[0]) - 1
        left = 0
        result = []
        while top <= bottom and left <= right:
            # top row
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1

            # right column
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1

            # bottom
            if top > bottom or left > right:
                break
            for i in range(right, left - 1, -1):
                result.append(matrix[bottom][i])
            bottom -= 1

            # left
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
            left += 1

        return result

File: Problem3_SpiralMatrix.cs
// Time Complexity : O(m * n)
// Space Complexity : O(m * n)
// Did this code successfully run on Leetcode : Yes

public class Solution {
    public IList<int> SpiralOrder(int[][] matrix) {
        int m = matrix.Length;
        int n = matrix[0].Length;
        int[] result = new int[m*n];
        int top = 0; int bottom = m - 1;
        int left = 0; int right = n - 1;
        int j = 0;
        while(top <= bottom && left <= right) {
            //top row
            for(int i = left; i <= right; i++) {
                result[j++] = matrix[top][i];
            }
            top++;
            //right column
            for(int i = top; i <= bottom; i++) {
                result[j++] = matrix[i][right];
            }
            right--;
            //bottom row
            if(top <= bottom) {
                for(int i = right; i >= left; i--) {
                    result[j++] = matrix[bottom][i];
                }
            }
            bottom--;
            //left column
            if(left <= right) {
                for(int i = bottom; i >= top; i--) {
                    result[j++] = matrix[i][left];
                }
            }
            left++;
        }
        return result;
    }
}
